src/plugins/ds_mongo.c

```diff
@@ -0,0 +1,3400 @@
+ * @return NULL on success;
+ * @return Sysrepo error info on error.
+ */
+sr_error_info_t *
+srpds_mongo_store(const struct lys_module *mod, sr_datastore_t ds, const struct lyd_node *mod_diff,
+        const struct lyd_node *mod_data, void *plg_data)
+{
+    mongo_data_t mdata;
+    mongo_plg_conn_data_t *pdata = (mongo_plg_conn_data_t *)plg_data;
+
+    sr_error_info_t *err_info = NULL;
+    int modified = 1;
+
+    assert(mod);
+
+    /* for candidate ds learn if modified */
+    if (ds == SR_DS_CANDIDATE) {
+        if ((err_info = srpds_mongo_candidate_modified(mod, plg_data, &modified))) {
+            return err_info;
+        }
+    }
+
+    /* if not modified, then copy running */
+    if (!modified) {
+        if ((err_info = srpds_mongo_copy(mod, SR_DS_CANDIDATE, SR_DS_RUNNING, plg_data))) {
+            return err_info;
+        }
+    }
+
+    if ((err_info = srpds_data_init(mod, ds, 1, pdata, &mdata))) {
+        goto cleanup;
+    }
+
+    if (ds == SR_DS_OPERATIONAL) {
+        if ((err_info = srpds_store_oper(mdata.module, mod_data))) {
+            goto cleanup;
+        }
+    }
+    else if (mod_diff) {
+        if ((err_info = srpds_store_all(mdata.module, mod_diff))) {
+            goto cleanup;
+        }
+    }
+
+    if ((err_info = srpds_set_last_modif_flag(mdata.module))) {
+        goto cleanup;
+    }
+
+    if (ds == SR_DS_CANDIDATE) {
+        if ((err_info = srpds_set_candidate_modified_flag(mdata.module, 1))) {
+            goto cleanup;
+        }
+    }
+    
+cleanup:
+    srpds_data_destroy(pdata, &mdata);
+    return err_info;
+}
+
+/**
+ * @brief Initialize data of a newly installed module.
+ *
+ * Init is called after fresh reboot of the system for every module for each datastore.
+ * Also, right after ::srds_install is called and afterwards the data of the module must be valid meaning any
+ * following ::srds_load __must return the stored data__ (may be empty).
+ *
+ * @param[in] mod Specific module.
+ * @param[in] ds Specific datastore.
+ * @param[in] plg_data Plugin data.
+ * @return NULL on success;
+ * @return Sysrepo error info on error.
+ */
+sr_error_info_t *
+srpds_mongo_init(const struct lys_module *mod, sr_datastore_t ds, void *plg_data)
+{
+    (void) plg_data;
+    (void) mod;
+    (void) ds;
+    return NULL;
+}
+
+/**
+ * @brief Initialize per-connection plugin data.
+ *
+ * If there is any module using the DS plugin, for each new connection this callback is called. Its purpose is
+ * to allow preparing per-process data. Since processes with running connections cannot be forked, the data will
+ * not be forked either.
+ *
+ * @param[in] conn New connection.
+ * @param[out] plg_data Arbitrary DS plugin data.
+ * @return NULL on success;
+ * @return Sysrepo error info on error.
+ */
+sr_error_info_t *
+srpds_mongo_conn_init(sr_conn_ctx_t *conn, void **plg_data)
+{
+    (void) conn;
+
+    sr_error_info_t *err_info = NULL;
+    mongo_plg_conn_data_t *data = NULL;
```

Comment:
Variable declarations should always be first.

Reply:
done